File: hosts/python/tabduct_host/native_messaging.py
"""Tabduct host — Chrome native-messaging transport (PROTOCOL.md §1).

Frames: ``[uint32 LE length][UTF-8 JSON]``. Read from stdin, write to stdout.
- A bad length header (<=0) or non-JSON body is a genuine desync → FATAL exit.
- An oversize frame (> MAX_FRAME_BYTES) is NOT a desync (the length is valid):
  we skip exactly that many bytes and keep the stream alive; the corresponding
  invoke reply is lost and times out. Mirrors hosts/node/src/native-messaging.js.

stdin is a blocking stream in Python, so the read loop runs on a dedicated
daemon thread and hands each decoded message to the asyncio loop via
``loop.call_soon_threadsafe``. Outbound writes (lifecycle replies + ``invoke``
requests) originate on the loop thread and are serialized under a lock.
"""

from __future__ import annotations

import asyncio
import json
import os
import sys
import threading
from typing import Any, Callable

from tabduct_host.constants import MAX_FRAME_BYTES, OUT_FRAME_MAX_BYTES


class FrameTooLargeError(Exception):
    """Raised when an outbound frame would exceed Chrome's 1 MB host->ext cap."""

    code = "FRAME_TOO_LARGE"


class NativeMessaging:
    def __init__(self, loop: asyncio.AbstractEventLoop) -> None:
        self._loop = loop
        self._on_message: Callable[[dict[str, Any]], None] | None = None
        self._on_end: Callable[[], None] | None = None
        self._out_lock = threading.Lock()
        self._thread = threading.Thread(target=self._read_loop, name="tabduct-nm-reader", daemon=True)
        self._stopped = False

    def on_message(self, fn: Callable[[dict[str, Any]], None]) -> None:
        self._on_message = fn

    def on_end(self, fn: Callable[[], None]) -> None:
        self._on_end = fn

    def start(self) -> None:
        self._thread.start()

    # --- inbound (reader thread) ------------------------------------------------

    def _read_loop(self) -> None:
        try:
            while True:
                header = self._read_exact(4)
                if header is None:
                    break  # stdin closed (extension gone / Chrome evicted worker)
                length = int.from_bytes(header, "little")
                if length <= 0:
                    self._fatal(f"bad length header: {length}")
                    return
                if length > MAX_FRAME_BYTES:
                    sys.stderr.write(
                        f"[tabduct] dropping oversize frame: {length}B (cap {MAX_FRAME_BYTES})\n"
                    )
                    sys.stderr.flush()
                    self._skip(length)
                    continue
                body = self._read_exact(length)
                if body is None:
                    break  # EOF mid-frame
                try:
                    msg = json.loads(body.decode("utf-8"))
                except Exception as e:  # non-JSON → unrecoverable desync
                    self._fatal(f"non-JSON frame: {e}")
                    return
                if not isinstance(msg, dict):
                    self._fatal(f"non-object frame: {type(msg).__name__}")
                    return
                self._loop.call_soon_threadsafe(self._dispatch, msg)
        except Exception as e:  # reader thread must never crash silently the wrong way
            sys.stderr.write(f"[tabduct] native-messaging reader error: {e}\n")
            sys.stderr.flush()
        finally:
            # Always signal end on the loop so the host shuts the server down.
            if not self._stopped:
                self._stopped = True
                try:
                    self._loop.call_soon_threadsafe(self._end)
                except RuntimeError:
                    pass  # loop already closed

    def _dispatch(self, msg: dict[str, Any]) -> None:
        if self._on_message is not None:
            self._on_message(msg)

    def _end(self) -> None:
        if self._on_end is not None:
            self._on_end()

    def _read_exact(self, n: int) -> bytes | None:
        """Read exactly ``n`` bytes. Returns None on EOF."""
        chunks: list[bytes] = []
        remaining = n
        while remaining > 0:
            chunk = sys.stdin.buffer.read(remaining)
            if not chunk:
                return None  # EOF
            chunks.append(chunk)
            remaining -= len(chunk)
        return b"".join(chunks)

    def _skip(self, n: int) -> None:
        remaining = n
        while remaining > 0:
            chunk = sys.stdin.buffer.read(min(remaining, 65536))
            if not chunk:
                return  # EOF while skipping
            remaining -= len(chunk)

    def _fatal(self, reason: str) -> None:
        sys.stderr.write(f"[tabduct] fatal framing error: {reason}\n")
        sys.stderr.flush()
        os._exit(1)  # length-prefixed desync is unrecoverable; Chrome surfaces the disconnect

    # --- outbound (loop thread) -------------------------------------------------

    def send(self, msg: dict[str, Any]) -> None:
        """Encode ``msg`` as a native-messaging frame and write it to stdout."""
        body = json.dumps(msg, separators=(",", ":")).encode("utf-8")
        if len(body) > OUT_FRAME_MAX_BYTES:
            raise FrameTooLargeError(
                f"outbound frame {len(body)}B exceeds 1 MB cap"
            )
        frame = len(body).to_bytes(4, "little") + body
        with self._out_lock:
            sys.stdout.buffer.write(frame)
            sys.stdout.buffer.flush()

File: hosts/dotnet/Tabduct.Host/NativeMessaging.cs
using System.Buffers.Binary;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Tabduct.Host;

/// <summary>
/// Chrome Native Messaging transport (PROTOCOL.md §1): frames are
/// <c>[uint32 LE length][UTF-8 JSON]</c>. stdin is read on a dedicated blocking
/// thread; stdout writes are serialized under a lock. Mirrors
/// hosts/python/tabduct_host/native_messaging.py + hosts/node/src/native-messaging.js.
/// </summary>
/// <remarks>
/// stdout MUST stay pure native-messaging frames: Kestrel/ASP.NET logging is kept
/// off stdout (see Program/McpServer), and this class writes raw bytes to the
/// standard-output stream captured once at construction — never Console.WriteLine.
/// </remarks>
internal sealed class NativeMessaging : IDisposable
{
    // Chrome caps: host->ext 1 MB (hard, sever-on-overflow); ext->host 64 MiB.
    // Our inbound sanity cap (about half the real inbound limit): oversize is
    // SKIPPED (the invoke times out), not fatal.
    private const int MaxFrameBytes = 32 * 1024 * 1024;
    private const int OutFrameMaxBytes = 1024 * 1024;

    private readonly Stream _stdin;
    private readonly Stream _stdout;
    private readonly object _writeLock = new();
    private Thread? _reader;
    private volatile bool _stopped;

    public NativeMessaging()
    {
        // Capture the raw streams ONCE. Writing bytes directly to the stdout
        // stream keeps it binary-clean (no TextWriter/CRLF/encoding translation).
        _stdin = Console.OpenStandardInput();
        _stdout = Console.OpenStandardOutput();
    }

    /// <summary>Wire error raised when an outbound frame would overflow Chrome's 1 MB cap.</summary>
    public sealed class FrameTooLargeException(string message) : Exception(message)
    {
        public string Code => "FRAME_TOO_LARGE";
    }

    /// <summary>Encode <paramref name="msg"/> as a native-messaging frame and write it to stdout.</summary>
    public void Send(JsonObject msg)
    {
        byte[] body = JsonSerializer.SerializeToUtf8Bytes(msg);
        if (body.Length > OutFrameMaxBytes)
        {
            throw new FrameTooLargeException($"outbound frame {body.Length}B exceeds 1 MB cap");
        }

        byte[] frame = new byte[4 + body.Length];
        BinaryPrimitives.WriteUInt32LittleEndian(frame.AsSpan(0, 4), (uint)body.Length);
        Buffer.BlockCopy(body, 0, frame, 4, body.Length);

        lock (_writeLock)
        {
            _stdout.Write(frame, 0, frame.Length);
            _stdout.Flush();
        }
    }

    /// <summary>Start the blocking stdin reader thread.</summary>
    /// <param name="onRequest">Invoked for inbound requests (have <c>type</c>).</param>
    /// <param name="onReply">Invoked for inbound replies (have <c>replyTo</c>); always called on the reader thread.</param>
    /// <param name="onEnd">Invoked once when stdin reaches EOF (authoritative shutdown).</param>
    public void Start(Action<JsonObject> onRequest, Action<JsonObject> onReply, Action onEnd)
    {
        _reader = new Thread(() => ReadLoop(onRequest, onReply, onEnd))
        {
            Name = "tabduct-nm-reader",
            IsBackground = true,
        };
        _reader.Start();
    }

    private void ReadLoop(Action<JsonObject> onRequest, Action<JsonObject> onReply, Action onEnd)
    {
        try
        {
            while (true)
            {
                byte[]? header = ReadExact(4);
                if (header is null) break; // stdin closed (extension gone / worker evicted)

                uint length = BinaryPrimitives.ReadUInt32LittleEndian(header);
                if (length == 0)
                {
                    Fatal($"bad length header: {length}");
                    return;
                }
                if (length > MaxFrameBytes)
                {
                    Log($"dropping oversize frame: {length}B (cap {MaxFrameBytes})");
                    Skip((int)length);
                    continue;
                }

                byte[]? body = ReadExact((int)length);
                if (body is null) break; // EOF mid-frame

                JsonObject? msg;
                try
                {
                    msg = JsonNode.Parse(Encoding.UTF8.GetString(body)) as JsonObject;
                }
                catch (Exception e)
                {
                    Fatal($"non-JSON frame: {e.Message}");
                    return;
                }
                if (msg is null)
                {
                    Fatal("non-object frame");
                    return;
                }

                if (msg.ContainsKey("replyTo"))
                {
                    onReply(msg);
                }
                else
                {
                    onRequest(msg);
                }
            }
        }
        catch (Exception e)
        {
            Log($"native-messaging reader error: {e.Message}");
        }
        finally
        {
            if (!_stopped)
            {
                _stopped = true;
                onEnd();
            }
        }
    }

    private byte[]? ReadExact(int n)
    {
        byte[] buf = new byte[n];
        int read = 0;
        while (read < n)
        {
            int r = _stdin.Read(buf, read, n - read);
            if (r <= 0) return null; // EOF (a partial frame at EOF is also treated as EOF)
            read += r;
        }
        return buf;
    }

    private void Skip(int n)
    {
        byte[] buf = new byte[Math.Min(n, 65536)];
        int remaining = n;
        while (remaining > 0)
        {
            int r = _stdin.Read(buf, 0, Math.Min(remaining, buf.Length));
            if (r <= 0) return; // EOF while skipping
            remaining -= r;
        }
    }

    private static void Fatal(string reason)
    {
        // Length-prefixed desync is unrecoverable: Chrome surfaces the disconnect.
        Log($"fatal framing error: {reason}");
        Environment.Exit(1);
    }

    private static void Log(string message)
    {
        // stderr only — never stdout (that is the frame stream).
        Console.Error.WriteLine($"[tabduct] {message}");
    }

    public void Dispose()
    {
        _stopped = true;
    }
}
